src/main/gpiiConnector.js

```diff
@@ -0,0 +1,280 @@
+ */
+gpii.app.createSettingModel = function (key, settingDescriptor) {
+    return {
+        path: key,
+        value: settingDescriptor.value,
+        solutionName: settingDescriptor.solutionName,
+
+        schema: settingDescriptor.schema,
+
+        // XXX hardcoded as they're not currently supported by the API (pcpChannel)
+        icon: fluid.module.resolvePath("%gpii-app/src/icons/gear-cloud-white.png"),
+        liveness: settingDescriptor.liveness || "live",
+        memory: fluid.isValue(settingDescriptor.memory) ? settingDescriptor.memory : true
+    };
+};
+
+/**
+ * Extracts data for the user's preference sets (including the active preference
+ * set and the applicable settings) from the message received when the user keys in
+ * or out.
+ * @param message {Object} The message sent when the user keys is or out (a JSON
+ * object).
+ * @return {Object} An object containing all preference sets, the active preference
+ * set and the corresponding settings.
+ */
+gpii.app.extractPreferencesData = function (message) {
+    var value = message.value || {},
+        preferences = value.preferences || {},
+        contexts = preferences.contexts,
+        settingControls = value.settingControls,
+        sets = [],
+        activeSet = value.activeContextName || null,
+        settings = [];
+
+    if (contexts) {
+        sets = fluid.hashToArray(contexts, "path");
+    }
+
+    if (settingControls) {
+        settings = fluid.values(
+            fluid.transform(settingControls, function (settingDescriptor, settingKey) {
+                return gpii.app.createSettingModel(settingKey, settingDescriptor);
+            })
+        );
+    }
+
+    return {
+        sets: sets,
+        activeSet: activeSet,
+        settings: settings
+    };
+};
+
+/**
+ * Extracts the user-friendly snapset name form the message received when the user
+ * keys in or out.
+ * @param message {Object} The message sent when the user keys is or out (a JSON
+ * object).
+ * @return {String} The user-friendly snapset name.
+ */
+gpii.app.extractSnapsetName = function (message) {
+    var value = message.value || {},
+        preferences = value.preferences || {};
+    return preferences.name;
+};
+
+/**
+ * Extension of `gpiiController` used for dev purposes.
+ */
+fluid.defaults("gpii.app.dev.gpiiConnector", {
+    gradeNames: ["gpii.app.gpiiConnector"],
+
+    events: {
+        mockPrefSets: {
+            event: "onPreferencesUpdated",
+            args: "@expand:gpii.app.dev.gpiiConnector.mockPreferences({arguments}.0)"
+        }
+    }
+});
+
+
+/**
+ * A decorator for the extracted preferences that applies values that are to be used
+ * for development.
+ */
+gpii.app.dev.gpiiConnector.mockPreferences = function (preferences) {
+    function applyManualLivenessFlag(settings) {
+        settings.forEach(function (setting) {
+            // XXX a workaround as the Magnifier settings are missing the `solutionName` property
```

Comment:
What will I do when I am no longer able to put such questions to @kaspermarkus ....